File: src/tools/renderer.cpp
/*
   Drawpile - a collaborative drawing program.

   Copyright (C) 2018-2019 Calle Laakkonen

   Drawpile is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   Drawpile is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with Drawpile.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "renderer.h"

#include "../libclient/canvas/statetracker.h"
#include "../libclient/canvas/layerlist.h"
#include "../libclient/canvas/aclfilter.h"
#include "../libclient/core/layerstack.h"
#include "../libclient/ora/orawriter.h"
#include "../libshared/record/reader.h"

#include <QImageWriter>
#include <QElapsedTimer>
#include <QPainter>

struct ExportState {
	QSize lastSize;
	int index;
};

QImage resizeImage(const QImage &img, const QSize &maxSize, bool fixedSize)
{
	if(fixedSize && img.size() != maxSize) {
		const QSize scaledSize = img.size().scaled(maxSize, Qt::KeepAspectRatio);
		QImage scaled(maxSize, QImage::Format_ARGB32_Premultiplied);
		scaled.fill(Qt::black);

		QPainter painter(&scaled);
		painter.setRenderHint(QPainter::SmoothPixmapTransform);
		painter.drawImage(
			QRect(
				QPoint((maxSize.width() - scaledSize.width()) / 2, (maxSize.height() - scaledSize.height()) / 2),
				scaledSize
			),
			img
			);
		return scaled;

	} else if (img.width() > maxSize.width() || img.height() > maxSize.height()) {
		return img.scaled(maxSize, Qt::KeepAspectRatio, Qt::SmoothTransformation);

	} else {
		return img;
	}
}

bool saveImage(const DrawpileCmdSettings &settings, const paintcore::LayerStack &layers, ExportState &state)
{
	if(layers.size().isEmpty() || layers.layerCount()==0) {
		// The layer stack has no size until the first resize command.
		// Trying to export before it is not a fatal error.
		if(settings.verbose)
			fprintf(stderr, "[I] Image is empty, not saving anything.\n");
		return true;
	}

	QString filename = settings.outputFilePattern;

	// Perform pattern subsitutions:
	// :idx: <-- image index number
	filename.replace(":idx:", QString::number(state.index));

	if(settings.verbose)
		fprintf(stderr, "[I] Writing %s...\n", qPrintable(filename));

	QString error;
	bool ok;
	if(filename.endsWith(".ora", Qt::CaseInsensitive)) {
		// Special case: Save as OpenRaster with all the layers intact
		// ORAs are not resized.
		ok = openraster::saveOpenRaster(filename, &layers, &error);

	} else {
		QImage flat = layers.toFlatImage(settings.mergeAnnotations, true);

		if(settings.fixedSize && state.lastSize.isEmpty())
			state.lastSize = flat.size();

		if(!state.lastSize.isEmpty())
			flat = resizeImage(flat, state.lastSize, settings.fixedSize);

		QImageWriter writer(filename);
		ok = writer.write(flat);
		if(!ok)
			error = writer.errorString();
	}

	if(!ok)
		fprintf(stderr, "[E] %s: %s\n", qPrintable(filename), qPrintable(error));

	++state.index;

	return ok;
}

QString prettyDuration(qint64 duration)
{
	const double msecs = duration / 1000000;
	if(msecs < 1000)
		return QStringLiteral("%1 ms").arg(msecs, 0, 'f', 1);
	const double secs = msecs / 1000;
	if(secs<60)
		return QStringLiteral("%1 s").arg(secs, 0, 'f', 2);
	return QStringLiteral("%1 m %2s").arg(secs/60, 0, 'f', 0).arg(fmod(secs, 60), 0, 'f', 2);
}

bool renderDrawpileRecording(const DrawpileCmdSettings &settings)
{
	// Open recording file
	recording::Reader reader(settings.inputFilename);
	recording::Compatibility compat = reader.open();

	if(compat == recording::CANNOT_READ) {
		fprintf(stderr, "[E] %s", qPrintable(reader.errorString()));
		return false;
	}

	if(compat != recording::COMPATIBLE && compat != recording::MINOR_INCOMPATIBILITY) {
		fprintf(stderr, "[E] Recording not compatible\n");
		return false;
	}

	// Initialize the paint engine
	paintcore::LayerStack image;
	canvas::LayerListModel layermodel;
	canvas::StateTracker statetracker(&image, &layermodel, 1);
	canvas::AclFilter aclfilter;

	aclfilter.reset(1, false);

	// Benchmarking
	QElapsedTimer renderTime;
	QElapsedTimer saveTime;
	QElapsedTimer totalTime;
	qint64 totalRenderTime = 0;
	qint64 totalSaveTime = 0;
	totalTime.start();

	// Prepare image exporter
	ExportState exportState {
		settings.maxSize,
		1
	};
	int exportCounter = 0;

	// Read and execute commands
	recording::MessageRecord record;
	do {
		const qint64 offset = reader.filePosition();
		record = reader.readNext();

		if(record.status == recording::MessageRecord::OK) {
			if(settings.acl && !aclfilter.filterMessage(*record.message)) {
				if(settings.verbose)
					fprintf(stderr, "[A] Filtered message %s from %d (idx %d @ %llx)",
						qPrintable(record.message->messageName()),
						record.message->contextId(),
						reader.currentIndex(),
						offset
						);
			}

			if(record.message->isCommand()) {
				renderTime.start();
				statetracker.receiveCommand(protocol::MessagePtr::fromNullable(record.message));
				totalRenderTime += renderTime.nsecsElapsed();
			}

			// Save images
			if(settings.exportEveryN > 0) {
				switch(settings.exportEveryMode) {
				case ExportEvery::Message:
					++exportCounter;
					break;
				case ExportEvery::Sequence:
					if(record.message->type() == protocol::MSG_UNDOPOINT)
						++exportCounter;
					break;
				}

				if(exportCounter >= settings.exportEveryN) {
					exportCounter = 0;
					saveTime.start();
					if(!saveImage(settings, image, exportState))
						return false;
					totalSaveTime += saveTime.nsecsElapsed();
				}
			}

		} else if(record.status == recording::MessageRecord::INVALID) {
			fprintf(stderr, "[E] Invalid message type %d at index %d, offset 0x%llx",
					record.invalid_type,
					reader.currentIndex(),
					offset
				   );
			return false;
		}
	} while(record.status != recording::MessageRecord::END_OF_RECORDING);

	fprintf(stderr, "[I] Total processing time: %s\n", qPrintable(prettyDuration(totalTime.nsecsElapsed())));
	fprintf(stderr, "[I] Cumulative render time: %s\n", qPrintable(prettyDuration(totalRenderTime)));

	// Save the final result
	saveTime.start();
	if(!saveImage(settings, image, exportState))
		return false;
	totalSaveTime += saveTime.nsecsElapsed();

	fprintf(stderr, "[I] Cumulative saving time: %s\n", qPrintable(prettyDuration(totalSaveTime)));

	return true;
}

File: src/drawpile-cli/src/renderer.rs
// This file is part of Drawpile.
// Copyright (C) 2020 Calle Laakkonen
//
// Drawpile is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// As additional permission under section 7, you are allowed to distribute
// the software through an app store, even if that store has restrictive
// terms and conditions that are incompatible with the GPL, provided that
// the source is also available under the GPL with or without this permission
// through a channel without those restrictive terms and conditions.
//
// Drawpile is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Drawpile.  If not, see <https://www.gnu.org/licenses/>.

use dpcore::canvas::CanvasState;
use dpcore::paint::LayerViewOptions;
use dpcore::protocol::message::{CommandMessage, Message};
use dpimpex::conv::from_dpimage;
use dpimpex::rec::reader::{open_recording, Compatibility, ReadMessage};

use tracing::{info, warn};

use std::error::Error;
use std::fmt;
use std::io;
use std::num::ParseIntError;
use std::path::Path;
use std::str::FromStr;
use std::time::{Duration, Instant};

use image;

#[derive(Clone, Copy, Eq, PartialEq)]
pub struct Size(u32, u32);

impl FromStr for Size {
    type Err = ParseIntError;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let x = s.find('x').unwrap_or(0);
        let w = s[..x].parse::<u32>()?;
        let h = s[x + 1..].parse::<u32>()?;
        Ok(Size(w, h))
    }
}

pub struct RenderOpts<'a> {
    /// Name of input recording file
    pub input_file: &'a str,

    /// Name of output image file (sequence number is inserted before the suffix)
    pub output_file: &'a str,

    /// How often to write the image
    pub output_every: Option<u32>,

    /// Whether to write a message or an undopoint
    pub every_up: bool,

    /// Resize images to this size
    pub resize: Option<Size>,

    /// Resize subsequent images to the original size
    pub same_size: bool,
}

struct RenderState {
    resize: Option<Size>,
    same_size: bool,
    image_num: u32,
}

#[derive(Debug)]
struct RenderError {
    message: &'static str,
}

impl fmt::Display for RenderError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl Error for RenderError {
    fn description(&self) -> &str {
        self.message
    }
}

pub fn render_recording(opts: &RenderOpts) -> Result<(), Box<dyn std::error::Error>> {
    let mut reader = open_recording(Path::new(opts.input_file))?;

    if reader.check_compatibility() == Compatibility::Incompatible {
        return Err(Box::new(RenderError {
            message: "Unsupported format version",
        }));
    }

    let start = Instant::now();
    let mut canvas = CanvasState::new();
    let mut total_render_time = Duration::new(0, 0);
    let mut total_save_time = Duration::new(0, 0);
    let mut message_counter = 0;
    let mut state = RenderState {
        resize: opts.resize,
        same_size: opts.same_size,
        image_num: opts.output_every.map_or(0, |_| 1),
    };

    loop {
        match reader.read_next() {
            ReadMessage::Ok(m) => {
                let now = Instant::now();
                if let Message::Command(c) = &m {
                    match (opts.every_up, c) {
                        (true, CommandMessage::UndoPoint(_)) | (false, _) => {
                            message_counter += 1;
                        }
                        _ => (),
                    }
                    canvas.receive_message(c);
                }
                total_render_time += now.elapsed();

                if let Some(e) = opts.output_every {
                    if message_counter >= e {
                        message_counter = 0;
                        total_save_time += save_canvas(opts, &mut state, &canvas)?;
                    }
                }
            }
            ReadMessage::Invalid(msg) => {
                warn!("Invalid message: {}", msg);
            }
            ReadMessage::IoError(e) => {
                return Err(Box::new(e));
            }
            ReadMessage::Eof => {
                break;
            }
        }
    }

    total_save_time += save_canvas(opts, &mut state, &canvas)?;

    let total_time = start.elapsed();

    info!(
        "Total render time: {:.3} s",
        total_render_time.as_secs_f64()
    );
    info!("Total save time: {:.3} s", total_save_time.as_secs_f64());
    info!("Total time: {:.3} s", total_time.as_secs_f64());

    Ok(())
}

fn save_canvas(
    opts: &RenderOpts,
    state: &mut RenderState,
    canvas: &CanvasState,
) -> io::Result<Duration> {
    let now = Instant::now();

    let filename = make_filename(opts, state.image_num);

    let mut img = from_dpimage(&canvas.layerstack().to_image(&LayerViewOptions::default()));

    let size = Size(img.width(), img.height());

    if let Some(resize) = state.resize {
        if size != resize {
            img = image::imageops::resize(
                &img,
                resize.0,
                resize.1,
                image::imageops::FilterType::CatmullRom,
            );
        }
    } else if state.same_size {
        state.resize = Some(size);
    }

    info!("Saving {}", filename);
    state.image_num += 1;

    match img.save(&filename) {
        Err(image::ImageError::IoError(e)) => Err(e),
        Err(e) => {
            warn!("Image encoding error: {}", e);
            Err(io::Error::new(io::ErrorKind::Other, "image encoding error"))
        }
        Ok(_) => Ok(now.elapsed()),
    }
}

fn make_filename(opts: &RenderOpts, index: u32) -> String {
    if opts.output_file.is_empty() {
        let end = opts.input_file.rfind('.').unwrap_or(opts.input_file.len());

        if index != 0 {
            format!("{}-{}.png", &opts.input_file[..end], index)
        } else {
            format!("{}.png", &opts.input_file[..end])
        }
    } else if index != 0 {
        let suffix = opts
            .output_file
            .rfind('.')
            .unwrap_or(opts.output_file.len());
        format!(
            "{}-{}{}",
            &opts.input_file[..suffix],
            index,
            &opts.output_file[suffix..]
        )
    } else {
        opts.output_file.to_string()
    }
}
